File: leetcode/python/316_remove_duplicate_letters.py
class Solution:
    def removeDuplicateLetters(self, s: str) -> str:
        stack: list[str] = []
        used = set[str]()
        last = {c: i for i, c in enumerate(s)}
        
        for i, c in enumerate(s):
            if c in used:
                continue
            while stack and c < stack[-1] and i < last[stack[-1]]:
                used.discard(stack.pop())
            used.add(c)
            stack.append(c)
        return ''.join(stack)

s = Solution()
print(s.removeDuplicateLetters("bcabc"))
print(s.removeDuplicateLetters("cbacdcbc"))

File: leetcode/rust/src/316_remove_duplicate_letters.rs
impl Solution {
    pub fn remove_duplicate_letters(s: String) -> String {
        let mut stack = vec![];
        let mut seen = vec![false; 26];
        let mut last_occurrence = vec![0; 26];
        for (i, c) in s.chars().enumerate() {
            last_occurrence[c as usize - 'a' as usize] = i;
        }

        for (i, c) in s.chars().enumerate() {
            let c = c as usize - 'a' as usize;
            if seen[c] {
                continue;
            }
            while let Some(&last) = stack.last() {
                if last > c && last_occurrence[last] > i {
                    seen[last] = false;
                    stack.pop();
                } else {
                    break;
                }
            }
            seen[c] = true;
            stack.push(c);
        }

        stack.iter().map(|&c| (c as u8 + 'a' as u8) as char).collect()
    }
}